### PR TITLE
feat(channels): Phase 7 — Channel Service Layer (#182)

### PR DESCRIPTION
**Pull request by dean0x**
## Problem Being Solved

Workers run as child processes (piped stdio), which will be metered starting June 15. Tmux sessions avoid this metering model and enable richer agent interaction.

## Key Changes

**EventDrivenWorkerPool** uses injected `TmuxConnectorPort`. Workers identified by tmux session name, not PID. Kill sequence is C-c → grace period → force-destroy. Output routing preserved through existing `OutputCapture` interface. Heartbeat detects dead sessions via `isAlive()`.

## Breaking Changes

- `WorkerRegistration.pid` is 0 for tmux workers
- `WorkerId` format changed from `worker-{pid}` to `worker-beat-{taskId}`
- `EventDrivenWorkerPoolDeps` interface changed (tmuxConnector + sessionsDir added)
- Requires tmux >= 3.0 installed

## Architecture

Incremental refactor of worker runtime to use tmux sessions instead of child processes.

### CODE REVIEW

**Comment by dean0x** [comment]
## Post-Resolution Review: Cycle 2 Analysis

**Status**: Cycle 2 (post-resolution review) of feat/182-channel-service-layer  
**PR**: #193 | **Branch**: feat/182-channel-service-layer → main  
**Date**: 2026-05-25 | **Reviewers**: 9 agents (Architecture, Performance, Complexity, Consistency, Reliability, Security, Testing, TypeScript, Regression)

---

### Summary Scores

| Category | Score | Status |
|----------|-------|--------|
| Architecture | 7/10 | CHANGES_REQUESTED |
| Performance | 7/10 | APPROVED_WITH_CONDITIONS |
| Complexity | 7/10 | APPROVED_WITH_CONDITIONS |
| Consistency | 8/10 | CHANGES_REQUESTED |
| Reliability | 8/10 | APPROVED_WITH_CONDITIONS |
| Security | 9/10 | APPROVED_WITH_CONDITIONS |
| Testing | 8/10 | APPROVED_WITH_CONDITIONS |
| TypeScript | 8/10 | APPROVED_WITH_CONDITIONS |
| Regression | 9/10 | APPROVED |

---

### Blocking Issues (≥80% Confidence)

#### 1. 🔴 Byte Length Check: content.length (UTF-16) vs Buffer.byteLength (UTF-8)
**File**: `src/implementations/tmux/tmux-session-manager.ts:379`  
**Confidence**: 85% | **Flagged by**: Security, Performance, Consistency, Reliability, TypeScript  
**Category**: MEDIUM - Correctness gap in defense-in-depth guard

The constant is documented as "Maximum byte length for paste content" and the error says "bytes", but uses `content.length` which counts UTF-16 code units, not bytes. Multi-byte characters (emoji, CJK) could produce ~1MB on disk vs 256KB stated limit.

**CLAUDE.md applies**: "byte-vs-char consistency" (2026-03-19) — totalSize measured in bytes via `Buffer.byteLength` everywhere.

**Fix**: `Buffer.byteLength(content, 'utf8') > MAX_PASTE_CONTENT_LENGTH`

---

#### 2. 🔴 Timer Leak: SerialQueue.drain() Never Clears setTimeout
**File**: `src/services/channel-manager.ts:85-88`  
**Confidence**: 85% | **Flagged by**: Performance, Complexity, Reliability  
**Category**: MEDIUM - Resource leak under sustained traffic

When `this.chain` resolves before timeout, the `setTimeout` at line 86 is never cleared. Each `sendMessage()` call at line 407 executes `drain(10_000)`, accumulating hundreds of orphaned timers over time. Prevents graceful shutdown and adds GC pressure.

**Fix**: 
```typescript
async drain(timeoutMs = 5_000): Promise<void> {
  let timer: ReturnType<typeof setTimeout>;
  const timeout = new Promise<void>((resolve) => {
    timer = setTimeout(resolve, timeoutMs);
  });
  await Promise.race([this.chain, timeout]);
  clearTimeout(timer!);  // ← Add this
}
```

---

#### 3. 🔴 Race Condition: sendMessage Returns Success When Drain Times Out
**File**: `src/services/channel-manager.ts:385-409`  
**Confidence**: 84% | **Flagged by**: Architecture, Performance, Reliability, TypeScript  
**Category**: MEDIUM - Data loss path (false-positive success)

`dispatchError` is mutated inside `queue.enqueue()` closure, then read after `drain(10_000)`. If drain timeout fires before the queued task executes (e.g., high load, slow preceding tasks), `dispatchError` remains `undefined`. Method returns `ok(undefined)` — a false-positive success. Caller has no indication the message was never delivered.

Also: type narrowing issue — `AutobeatError` → `Error` loses error code.

**Fix**: Use Promise to capture result reliably instead of mutable closure across async boundary.

---

#### 4. 🔴 Architecture Inconsistency: Constructor Subscriptions Bypass Factory Pattern
**File**: `src/services/channel-manager.ts:155-175`  
**Confidence**: 85% | **Flagged by**: Architecture  
**Category**: HIGH - Consistency with established patterns

ChannelManager subscribes to `ChannelDestroyed` and `ChannelMemberCrashed` events directly in the constructor, silently ignoring subscription errors. Every other handler in the codebase (ChannelHandler, ScheduleHandler, LoopHandler) uses a factory pattern:

```typescript
static async create(deps): Result<Handler> { ... }
private subscribeToEvents(): Result<void> { ... }
```

This allows subscription failures to surface to the caller rather than being silently swallowed.

**Impact**: If `eventBus.subscribe` fails, ChannelManager proceeds without the ChannelDestroyed subscription. Event-driven destroy paths (max-rounds-reached, all-members-crashed) would never trigger session teardown or DB status update.

**Fix**: Extract subscriptions to factory pattern matching ChannelHandler, ScheduleHandler, and other handlers.

---

### Should-Fix Issues (80%+ Confidence, Not Blocking)

#### 5. 📋 Interface Gap: recoverChannels() and dispose() Not on ChannelService Interface
**File**: `src/core/interfaces.ts:1054-1081` | `src/bootstrap.ts:757` | `src/index.ts:96`  
**Confidence**: 82% | **Flagged by**: Architecture  
**Category**: HIGH - DIP violation (concrete type casts at call sites)

`recoverChannels()` and `dispose()` are public methods on ChannelManager but absent from the ChannelService interface. Bootstrap and shutdown code must cast ChannelService to concrete ChannelManager type or use structural type assertions. This breaks the Dependency Inversion Principle.

**Fix**: Add lifecycle methods to ChannelService interface (matching ScheduleService, LoopService pattern):
```typescript
export interface ChannelService {
  // ... existing methods ...
  recoverChannels(): Promise<Result<void>>;
  dispose(): void;
}
```

---

#### 6. 📋 Channel Cache Staleness: currentRound Never Refreshed on Round Increment
**File**: `src/services/channel-manager.ts:522, 883-889`  
**Confidence**: 82% | **Flagged by**: Architecture  
**Category**: MEDIUM - Potential for downstream issues if cache used for decisions

The `channelCache` stores Channel objects at create/recover time but is only invalidated on member crash and channel destroy. When ChannelHandler increments `currentRound` via DB, there is no cache invalidation. The cache entry's `currentRound` can lag indefinitely (comment at line 140 acknowledges this but bounds are unclear).

**Impact**: If cache is later used for round-based decisions (not just informational payloads), stale data could cause incorrect behavior. Currently used for event payloads only (acceptable risk), but worth bounding.

---

#### 7. 📋 Logging Inconsistency: handleMemberExitAsync Uses warn() Instead of error()
**File**: `src/services/channel-manager.ts:948-953`  
**Confidence**: 92% | **Flagged by**: Consistency  
**Category**: HIGH - Pattern violation (established logger usage)

The `handleMemberExitAsync` catch block uses `logger.warn(msg, {error: string})` while every other emit failure in this file uses `logger.error(msg, Error, context)` (3-argument pattern). The prior resolution cycle fixed 5 locations to use this pattern; this catch block was not addressed because it's in a Promise chain rather than an if-guard.

**Fix**: Align to 3-argument pattern:
```typescript
.catch((e: unknown) => {
  const error = e instanceof Error ? e : new Error(String(e));
  this.logger.error('Failed to emit ChannelMemberCrashed', error, {
    channelId,
    memberName,
  });
});
```

---

#### 8. 📋 Logging Inconsistency: Constructor ChannelDestroyed Subscription Uses warn()
**File**: `src/services/channel-manager.ts:161-166`  
**Confidence**: 90% | **Flagged by**: Consistency  
**Category**: HIGH - Pattern violation

Same pattern issue as #7 — uses `logger.warn(msg, {error: string})` instead of `logger.error(msg, Error, context)`.

**Fix**: Align to 3-argument pattern.

---

#### 9. 📋 Unchecked Emit: recoverSingleChannel Emits ChannelDestroyed Without Checking Result
**File**: `src/services/channel-manager.ts:509`  
**Confidence**: 82% | **Flagged by**: Consistency  
**Category**: MEDIUM - Missing error handling

When all members are dead, `recoverSingleChannel` calls `await this.eventBus.emit('ChannelDestroyed')` without checking the result. Every other emit call in this file checks `emitResult.ok` and either returns error or logs it. The sole exception is this recovery path.

**Fix**: Log emit failures for observability (consistent with established pattern).

---

#### 10. 📋 Iterate-and-Delete Anti-Pattern: cleanupInMemory Deletes During Iteration
**File**: `src/services/channel-manager.ts:1025-1029`  
**Confidence**: 82% | **Flagged by**: Complexity  
**Category**: MEDIUM - Code smell (iteration safety)

The `for...of` loop over `this.memberHandles` calls `delete(key)` inside the iteration body. While technically safe in ES6, this is a well-known code smell that confuses readers and can silently break if logic is modified. Also couples to `handleKey` format via string-prefix check.

**Fix**: Collect keys first, delete in second pass:
```typescript
const keysToDelete: string[] = [];
for (const [key, handle] of this.memberHandles) {
  if (key.startsWith(prefix)) {
    this.sessionToChannel.delete(handle.sessionName);
    keysToDelete.push(key);
  }
}
for (const key of keysToDelete) {
  this.memberHandles.delete(key);
}
```

---

### Testing Gaps (Blocking)

#### 11. 🧪 Mock Incomplete: batchUpdateMemberStatuses Missing from channel-handler.test.ts Mock
**File**: `tests/unit/services/handlers/channel-handler.test.ts:52-86`  
**Confidence**: 85% | **Category**: MEDIUM - Interface contract violation

The `makeChannelRepo()` mock declares return type `ChannelRepository` but does not include the newly-added `batchUpdateMemberStatuses` method. This creates an incomplete mock. Future changes to ChannelHandler calling this method would crash in tests.

**Fix**: Add method to mock:
```typescript
batchUpdateMemberStatuses: vi
  .fn()
  .mockImplementation(async (id, memberNames, status) => {
    const ch = channels.get(id);
    if (ch) {
      const nameSet = new Set(memberNames);
      channels.set(id, {
        ...ch,
        members: ch.members.map((m) => 
          nameSet.has(m.name) ? { ...m, status } : m
        ),
      });
    }
    return ok(undefined);
  }),
```

---

#### 12. 🧪 Missing Boundary Test: Zero Members Validation Not Tested
**File**: `tests/unit/services/channel-manager.test.ts`  
**Confidence**: 82% | **Category**: MEDIUM - Untested validation

`validateCreateRequest()` at line 596 rejects `members.length === 0`, but no test exercises this path. Tests cover `>10` members and duplicates but not the empty array boundary.

**Fix**: Add test case for zero members.

---

### Lower-Confidence Suggestions (60-79%)

These did NOT get inline comments per your guidance (only ≥80% confidence inline):

- **SerialQueue drain timeout is a magic number** (65%) — Extract 5000ms and 10000ms to named constants
- **Parallel spawn with Promise.allSettled swallows individual errors** (70%) — Only `firstError` returned; others silently lost
- **broadcastToActiveMembers delivers sequentially** (82%) — Could parallelize member delivery within single routing task (should-fix, not blocking)
- **deliverTopic sequential delivery** (70%) — One-time operation, not a hot path
- **Two DB reads in handleChannelDestroyedEvent** (62%) — Could check cache first
- **channelCache stale round data** (65%) — Documented as acceptable but document bounds explicitly
- **No test for handleChannelDestroyedEvent when initiated by ChannelHandler** (82%) — Should-fix coverage gap
- **Paused channel recovery not tested** (80%) — Should-fix coverage gap
- **No test for routeAndDeliverMessage with cache miss** (65%) — Low priority
- **No test for sendMessage when channel has no queue** (65%) — Defensive test

---

### Overall Assessment

**Prior Resolution Summary**: 37 total issues from cycle 1 → 36 fixed, 1 false positive. Key fixes confirmed: dispose() wired, ChannelDestroyed handled, factory pattern applied to most handlers, tests strengthened.

**This Cycle**: 12 blocking/should-fix issues identified, categorized by severity and impact. 9 reviewers achieved strong consensus on the most critical items:

1. **byte-length inconsistency** — applies documented project decision
2. **timer leak** — resource accumulation under sustained load  
3. **dispatch race** — false-positive success, data loss path
4. **constructor subscriptions** — consistency with factory pattern
5. **interface gaps** — DIP violations with concrete casts
6. **logging inconsistencies** — pattern deviations (2 sites)

**Recommendation**: **CHANGES_REQUESTED**

This PR represents solid architectural progress (clean decomposition, event-driven patterns, performance optimizations). All findings are actionable with clear fixes. None require architectural redesign; all are localized to specific methods/patterns.

**Next Steps**: Apply fixes to the 12 items above (prioritize #1-6 as blocking, #7-10 as should-fix). Re-run affected test suites post-fix. Consider extracting `InMemoryChannelState` class in a follow-up to address file-level complexity (1047 lines approaches critical threshold).

---

**Review Session**: Post-resolution cycle 2 | **Date**: 2026-05-25 | **PR**: #193

File: src/services/handlers/channel-handler.ts [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+      this.logger.info('Channel round incremented', { channelId, round: newRound });
+
+      // Check termination bound
+      if (maxRounds !== undefined && newRound >= maxRounds) {
+        this.logger.info('Channel reached maxRounds — destroying', { channelId, maxRounds });
+        const destroyResult = await this.eventBus.emit('ChannelDestroyed', {
+          channelId,
+          reason: 'max-rounds-reached',
+        });
+        if (!destroyResult.ok) return destroyResult;
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/services/handlers/channel-handler.ts:156]: Already handled — `ChannelManager` subscribes to `ChannelDestroyed` at line 188 (`subscribeToEvents()`) and tears down sessions, updates DB status, and cleans in-memory state in `handleChannelDestroyedEvent()` (lines 1070-1104).

File: src/services/handlers/channel-handler.ts [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+ * Round counting rules:
+ *   - broadcast/directed: a round is complete when all active members have spoken
+ *     since the last round increment. The participation set is reset on each increment.
+ *   - round-robin: a round is complete when the turn cycles back to the first member.
+ *     The handler relies on the from field in ChannelMessageSent events.
+ *   - external messages (from === 'external') do not count as member turns.
+ */
+
+import type { ChannelId } from '../../core/domain.js';
+import { ChannelMemberStatus } from '../../core/domain.js';
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/services/handlers/channel-handler.ts]: Fixed — imported `ChannelStatus` alongside `ChannelMemberStatus`.

File: src/services/handlers/channel-handler.ts [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    await this.handleEvent(event, async (e) => {
+      const { channelId, from } = e;
+
+      // External messages do not count as member turns
+      if (from === 'external') return ok(undefined);
+
+      const channelResult = await this.channelRepository.findById(channelId);
+      if (!channelResult.ok) return channelResult;
+      const channel = channelResult.value;
+      if (!channel || channel.status !== 'active') return ok(undefined);
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/services/handlers/channel-handler.ts]: Fixed — replaced bare `active` string with `ChannelStatus.ACTIVE`.

File: src/services/channel-manager.ts [hunk 4]
```diff
@@ -0,0 +1,1158 @@
...
+      name: channel.name,
+      members: channel.members.map((m) => m.name),
+      communicationMode: channel.communicationMode,
+    });
+    if (!emitResult.ok) {
+      this.logger.error('Failed to emit ChannelCreated event', emitResult.error, { channelId: channel.id });
+      await this.destroyHandles(spawnedHandles.map((s) => s.handle));
+      this.cleanupInMemory(channel.id);
+      return err(emitResult.error);
+    }
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/services/channel-manager.ts:270]: Fixed — added `channelRepository.delete(channel.id)` to the error path after `cleanupInMemory`, with a warn log if the delete itself fails.

File: src/services/handlers/channel-handler.ts [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+    const seenFirst = this.rrFirstMemberSeen.get(channelId) ?? false;
+    if (seenFirst) {
+      // The first member has spoken again → full cycle completed
+      return true;
+    }
+
+    // First time we see the first member — record it but don't increment yet
+    this.rrFirstMemberSeen.set(channelId, true);
+    return false;
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/services/handlers/channel-handler.ts:195]: Fixed — added `this.rrFirstMemberSeen.set(channelId, false)` when `checkRoundRobinComplete` returns `true`. Also added a regression test that verifies A speaking first in a new cycle marks (not completes) the round, and updated the existing consecutive-rounds test to send A twice per cycle (matching the corrected semantics).

File: src/services/channel-manager.ts [hunk 6]
```diff
@@ -0,0 +1,1165 @@
...
+    const routeResult = ChannelRouter.route(channel, memberName, directedTo);
+    if (!routeResult.ok) {
+      this.logger.debug('No routing targets for member output', {
+        channelId,
+        memberName,
+        error: routeResult.error.message,
+      });
+      return;
+    }
+
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/services/channel-manager.ts:1018]: Pre-existing architectural limitation — `TmuxConnectorPort` has no `attach()` or `monitor()` method to re-register `onOutput`/`onExit` callbacks on recovered sessions. Adding callback re-attachment requires a new port method + connector implementation, which is out of scope for this PR. Filed as a follow-up concern — recovery currently restores routing state (handles, maps, delivery) but not output capture.
